### DISCUSSION TITLE
Preview 1.10.4_01 with "jump in" and "jump out".

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#482] [Announcements]
Hello folks,

preview 1.10.4_01 adds menu items View->Jump in and View->Jump out as suggested in #455 "View node as a sub-map". If you can find better names, please let us know. 

The implementation had to touch some core parts of the Freeplane and might have side effects. As always I am thankful for your feedback and bug reports.

This  preview also extends the style editor tool bar by new buttons for style management. It uses new icons created by @PredragC.

Cheers,
Dimitry

**Comment by euu2021** [comment]
Nice!! I'm eager to test it, but I will only be able to do it next weekend.

**Reply by Illegit** [reply]
NIce, it is working well, it suits my needs. Very appreciated.
Thank you
illegit
PS; just a note about: when you move out of the subtree without using jump-out it generates an error, is it avoidable? But in any case with jump-out it is running very well again.

**Reply by dpolivaev** [reply]
Thanks, I 'll  fix it

**Comment by abc16361** [comment]
When I press Esc (Goto root), it turns out the same result as Jump out. Is that what we expect?

**Reply by dpolivaev** [reply]
Probably not, I 'll  fix it

**Comment by Illegit** [comment]
Gradually using it I have something to signal
another strange behavior:
when you jump-in a node and after you want to move with the cursor using the arrows, the node remains selected, and to move with the selection you have to push the key "ESC" after the arrows. Is it expected?
illegit

**Reply by dpolivaev** [reply]
I am already building the new version 1.10.4_02. Please check if the problem is still relevant there.

**Comment by dpolivaev** [comment]
Preview 1.10.4_02 fixing all reported bugs is available

**Comment by Illegit** [comment]
Hi,  with FP 1.10.4_02 the previous errors disappeared. There is a behavior, expected?. when you jump-in or jump-out in a node to move out of that node with the arrows you need to push once the esc key.
In any case is working pretty well.
illegit

**Reply by dpolivaev** [reply]
Thanks, I'll fix it

**Reply by dpolivaev** [reply]
Fixed in preview 1.10.4_03

**Reply by Illegit** [reply]
Thank you Dimitry, confirmed

**Reply by awolo** [reply]
Problems when moving left out and up to the real, temporarily invisible, parent. Even with very simple new map with 3 or 4 plain vanilla nodes, but also with saved maps.
Various crashes, sometimes with popup trace, sometimes with error report in staus bar - up to a dozen errors sometimes. Sometimes long pause, sometimes bad redraw problems.
![FP jump crash](https://user-images.githubusercontent.com/5672151/178549307-9ba70c5f-1b09-47ce-814b-1ec3ff1ccff0.jpg)

**Reply by dpolivaev** [reply]
@awolo please attach a zipped log file, an example mind map, and a detailed description of what you do.

**Reply by dpolivaev** [reply]
@awolo If nobody knows how to trigger the bug and nobody shares the map where it can be observed no bug fix is possible.
If you are so lucky to hunt a bug please provide better evidence.

**Reply by awolo** [reply]
Please try this example. It is quick to reproduce, and it generates an unexpected condition. But there may be no log file. I've included a log file to an earlier test.

An automatic Jump Out should happen, but instead the "~ New Map ~" appears and the screen is unresponsive. In this case there is no Automatic Bug Report. After a minute or so the "!" triangle may (sometimes!) appears with text "13 errors" in the status line. This last time it showed 20 errors, but the file closed with no bug report.

I have sent you a few of the earlier Automatic Bug Reports already, but these simple examples are not generating them.

|file |new map
Template: standard-1.6.mm
insert 3 children to root: a, b, and c [I added a few more since]
save map
 [all moves made with |Edit |Move and sort |Move node to the left] 
move b to be child of a, c to be child of b
save map
"Jump in" with focus on a
promote b from child of a  -- THIS GENERATES ERROR CONDITION --
I see no unusual CPU or disk activity, but the screen does not initially respond.
[log file and mm.zip](https://github.com/freeplane/freeplane/files/9097548/log.file.and.mm.zip)

**Reply by dpolivaev** [reply]
Excellent description, I can see it.

**Reply by awolo** [reply]
Thanks. Glad I could please you! 
You please us constantly, so it's nice to be able to give back to you.

**Reply by dpolivaev** [reply]
Fixed in preview 1.10.4_05

**Comment by ldmpub** [comment]
Hi all,

**Is it possible to have a demo video of this new feature in action?**
(maybe it already exists in the documentation, if so could you please point me to it? because I did not find it).

Why this request? I want to translate this feature in french, maybe help to find a better "noun" for it _(see @dpolivaev request in first post above)_ and both without installing Freeplane preview because a stable version of Freeplane is required for my daily tasks 😉

Thanks,

**Comment by EdoFro** [comment]
> If you can find better names, please let us know.

what about using the concept "Focus" instead of "Jump"

Something like:

- "Set focus" / "Remove focus"
- "focus view" / "general view"
- "focus view" / "broad view"
- "Focus" / "Unfocus"
- "Focus in" / "Focus out" ¿?

**Reply by awolo** [reply]
WikidPad uses "Set as root". I would like "Set as temporary root" but it's a little long.
Also, it's kind of "Spotlight with Children"

**Reply by dpolivaev** [reply]
"Jump in" can be explained as "Set as view root" and "Jump out" can be explained as "Reset view root".
I think that "jump in" and "jump out" is shorter and more descriptive, but I think the explaining sentences could be used as their tool tips.
How do you see it?
@ldmpub @awolo

**Reply by awolo** [reply]
I like using tool tips to clarify.
"Temporary root: jump in" and "Temporary root: jump out"
Actually, for tool tips if you just use "Set root" and "Reset root" I think people would get it.

**Reply by EdoFro** [reply]
and "Local view" / "Global view" ?

or "Show as root"

"Set as root" sounds like making a change to the map, not?

**Reply by abc16361** [reply]
In some text editors, they use ' Hoist ' to display only the selected text.

**Reply by ldmpub** [reply]
Hi,

I have read your proposals and have yet decided to use "**View as root node**" for this feature in the french translation.

Remarks
- feature name is quite long (specially in french!) ... but I can find any shorter one
- a tooltip would help but I'm not sure it's a good practice to add tooltip in a menu (it sounds like popup above popup)
- I hope this feature will be added / grouped in the "Display" menu (I haven't installed FP preview yet, nor I have seen any documentation / video of this new feature).

Regards,

**Reply by dpolivaev** [reply]
I have added the tool tips as follows:
* JumpInAction.text=Jump in
* JumpInAction.tooltip=Change view root node
* JumpOutAction.text=Jump out
* JumpOutAction.tooltip=Reset view root node

**Reply by dpolivaev** [reply]
If "Change view root node" were used without "Jump in" one could expect that ancestors of this node would be displayed as a new subtree. The use of this text as a tool tip in the context of "Jump in" seems to be clear enough.

**Reply by wordmuse** [reply]
In various outliner programs, this is called "hoist" and "unhoist."  But I'm easy. 

~ Bal

**Comment by euu2021** [comment]
There is something that is still not clear to me: is "left side expansion" planned to be implemented? 

By "left side expansion" I mean: the nodes growing to the left side of the submap root, like happens with the left side of the real Root. It seems that this was the core of the #455 request, wasn't it?

**Reply by dpolivaev** [reply]
In my eyes, the core of the request is already implemented.
Your idea https://github.com/freeplane/freeplane/discussions/455#discussioncomment-2951222 to have the nodes shown at the different sizes is doable, but I am not sure if it is more helpful or more confusing if some nodes change their sides after the jump in.

Please start a separate discussion so that I can see which feedback comes specifically about this suggestion.

**Reply by ldmpub** [reply]
If nodes were changing of sides around the root node it would be more than "just a jump" but also a "node reorganization" ; a behavior I won't except as an end-user but I'm eager to ear pro arguments ...

**Reply by euu2021** [reply]
If this is implemented, maybe another name for the feature will make it more clear. 

Maybe: "View as Submap". If it's about a "Submap", then the user will expect to see left side expansion, because that's how maps behave.

**Comment by grotesque** [comment]
Hi,
Pressing Escape/esc brings freeplane out of the jump. Is this expected? Can this be disabled and can we let jump-out be the only way to get out of the jump? I've a habit of spamming escape/esc key.

**Reply by macmarrum** [reply]
By default, Escape is a hot key for `Navigate->Goto root`. And any navigation to a node outside of a jump-in view does a jump out.

You might want to remove or change the key binding. Just go to `Navigate` and `ctr+click` on `Goto root`.

**Comment by dpolivaev** [comment]
Escape on root triggers jumping out. Does everyone agree it shouldn't be like that?

**Reply by macmarrum** [reply]
Then it works differently than I thought.

I don't have an opinion. It makes sense both ways.

**Reply by euu2021** [reply]
I agree. Giving more nuances to the Esc key makes the UX too complicated. I think that it gets easier to interact when, as a general rule, each key has only one function on any context.

**Reply by grotesque** [reply]
I approach the issue from psychological perspective. I think jump inside a node is used to forget everything else and concentrate on that node and its descendants without getting "disturbed". Pressing esc/escape key, a key that's very important in freeplane owning to its existences as one of the default shortcuts, taking me out and "disturbing" me defeats the purpose of the jump. 

Maybe we could disable go-to to everything outside the jump-in view like go-backwards(alt+left) etc.

**Reply by wordmuse** [reply]
Before this, Esc took me to the mindmap's root. I would thus prefer that, when jumped in, it should take me to the new visible root.  Jump out should be its own command.  

I do like this Jump-in, Jump-out feature. Very slick.

~ Bal
